File: notes-swift/Sources/Notes/Sync/SyncService.swift
import Foundation

@MainActor
final class SyncService: ObservableObject {
    @Published private(set) var isSyncing = false

    private let api = APIClient.shared
    private let state = SyncStateStore.shared
    private let store: AppStore

    init(store: AppStore = .shared) {
        self.store = store
    }

    func syncNow() async {
        guard !isSyncing else { return }
        isSyncing = true
        defer { isSyncing = false }

        do {
            let cursor = state.cursor
            let ops = buildOps(since: cursor)
            let request = SyncRequest(cursor: cursor, ops: ops)
            let response: SyncResponse = try await api.post("/sync", body: request)
            applyServerChanges(response)
            state.cursor = response.nextCursor
        } catch {
            print("[sync] failed: \(error)")
        }
    }

    // MARK: - Build outbound ops

    private func buildOps(since cursorString: String?) -> [SyncOperation] {
        let cursor = cursorString.flatMap { ISO8601DateFormatter().date(from: $0) }
        let iso = ISO8601DateFormatter()
        var ops: [SyncOperation] = []

        for note in store.notes {
            guard cursor == nil || note.updatedAt > cursor! else { continue }
            let updatedAt = iso.string(from: note.updatedAt)
            if note.deletedAt != nil {
                ops.append(SyncOperation(id: UUID().uuidString, type: "delete", entityType: "note", entityId: note.id, updatedAt: updatedAt, payload: nil))
            } else {
                ops.append(SyncOperation(id: UUID().uuidString, type: "upsert", entityType: "note", entityId: note.id, updatedAt: updatedAt, payload: [
                    "title": AnyCodable(value: note.title),
                    "content": AnyCodable(value: note.content),
                    "isPinned": AnyCodable(value: note.isPinned),
                    "folderId": AnyCodable(value: note.folderId as Any),
                ]))
            }
        }

        for folder in store.folders {
            guard cursor == nil || folder.updatedAt > cursor! else { continue }
            ops.append(SyncOperation(id: UUID().uuidString, type: "upsert", entityType: "folder", entityId: folder.id, updatedAt: iso.string(from: folder.updatedAt), payload: [
                "name": AnyCodable(value: folder.name),
            ]))
        }

        for task in store.tasks {
            guard cursor == nil || task.updatedAt > cursor! else { continue }
            let updatedAt = iso.string(from: task.updatedAt)
            if task.deletedAt != nil {
                ops.append(SyncOperation(id: UUID().uuidString, type: "delete", entityType: "task", entityId: task.id, updatedAt: updatedAt, payload: nil))
            } else {
                ops.append(SyncOperation(id: UUID().uuidString, type: "upsert", entityType: "task", entityId: task.id, updatedAt: updatedAt, payload: [
                    "title": AnyCodable(value: task.title),
                    "description": AnyCodable(value: task.taskDescription),
                    "isCompleted": AnyCodable(value: task.isCompleted),
                ]))
            }
        }

        return ops
    }

    // MARK: - Apply server changes

    private func applyServerChanges(_ response: SyncResponse) {
        let iso = ISO8601DateFormatter()

        for tombstone in response.deleted {
            if tombstone.entityType == "note",
               let idx = store.notes.firstIndex(where: { $0.id == tombstone.entityId }) {
                var note = store.notes[idx]
                note.deletedAt = iso.date(from: tombstone.deletedAt)
                store.upsertNote(note)
            }
        }

        for dto in response.notes {
            let serverUpdatedAt = iso.date(from: dto.updatedAt) ?? Date()
            let note = Note(id: dto.id, userId: dto.userId, folderId: dto.folderId,
                            title: dto.title, content: dto.content, isPinned: dto.isPinned,
                            createdAt: iso.date(from: dto.createdAt) ?? Date(),
                            updatedAt: serverUpdatedAt)
            if let existing = store.notes.first(where: { $0.id == dto.id }),
               existing.updatedAt >= serverUpdatedAt { continue }
            store.upsertNote(note)
        }

        for dto in response.folders {
            let serverUpdatedAt = iso.date(from: dto.updatedAt) ?? Date()
            if let existing = store.folders.first(where: { $0.id == dto.id }),
               existing.updatedAt >= serverUpdatedAt { continue }
            store.upsertFolder(Folder(id: dto.id, userId: dto.userId, name: dto.name,
                                     createdAt: iso.date(from: dto.createdAt) ?? Date(),
                                     updatedAt: serverUpdatedAt))
        }

        for dto in response.tasks {
            let serverUpdatedAt = iso.date(from: dto.updatedAt) ?? Date()
            if let existing = store.tasks.first(where: { $0.id == dto.id }),
               existing.updatedAt >= serverUpdatedAt { continue }
            store.upsertTask(NoteTask(id: dto.id, userId: dto.userId, title: dto.title,
                                     taskDescription: dto.description, isCompleted: dto.isCompleted,
                                     createdAt: iso.date(from: dto.createdAt) ?? Date(),
                                     updatedAt: serverUpdatedAt))
        }
    }
}

File: lib/services/sync_service.dart
import 'dart:convert';
import 'dart:developer' as dev;
import 'package:drift/drift.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../database/database.dart';
import '../database/daos.dart';
import '../utils/note.dart';
import 'api_service.dart';

class SyncService {
  final AppDatabase db;
  final ApiService api;
  final SyncOpDao _opDao;
  static const String _cursorKey = 'sync_cursor';

  SyncService(this.db, this.api, this._opDao);

  /// Pushes queued local mutations to the server, then pulls remote changes.
  ///
  /// Op shape matches server/types/sync.ts:
  ///   {id, type: 'upsert'|'delete', entityType, entityId, updatedAt, payload}
  /// Server replies with:
  ///   {nextCursor, notes, folders, deleted, processedOpIds, errors?}
  Future<void> syncData(String userId) async {
    final prefs = await SharedPreferences.getInstance();
    final cursor = prefs.getString(_cursorKey);

    // 1. Drain pending ops FIFO.
    final pending = await _opDao.pullPending();
    final ops = pending
        .map((op) => {
              'id': op.id,
              'type': op.opType,
              'entityType': op.entityType,
              'entityId': op.entityId,
              'updatedAt': op.updatedAt.toIso8601String(),
              'payload': jsonDecode(op.payload),
            })
        .toList();

    try {
      final response = await api.post('sync', {
        'cursor': cursor,
        'ops': ops,
      });

      if (response.statusCode != 200) {
        dev.log('Sync failed: ${response.statusCode}');
        return;
      }

      final data = response.data as Map<String, dynamic>? ?? {};

      // 2. Server tells us which ops it applied — only clear those.
      final processedIds = (data['processedOpIds'] as List<dynamic>? ?? const [])
          .whereType<String>()
          .toList();
      if (processedIds.isNotEmpty) {
        await _opDao.deleteOps(processedIds);
      }

      final errors = data['errors'] as List<dynamic>?;
      if (errors != null && errors.isNotEmpty) {
        dev.log('Sync: ${errors.length} ops rejected by server: $errors');
      }

      // 3. Persist the new cursor.
      final nextCursor = data['nextCursor'] as String?;
      if (nextCursor != null) {
        await prefs.setString(_cursorKey, nextCursor);
      }

      // 4. Apply pulled data (folders first so FKs exist for notes).
      final foldersList = data['folders'] as List<dynamic>? ?? const [];
      for (final f in foldersList) {
        if (f is! Map) continue;
        final m = Map<String, dynamic>.from(f);
        final id = m['id'] as String?;
        final name = m['name'] as String?;
        final uid = m['userId'] as String?;
        if (id == null || name == null || uid == null) continue;
        await db.into(db.folders).insertOnConflictUpdate(
          FoldersCompanion.insert(
            id: id,
            name: name,
            userId: uid,
            deletedAt: Value(_parseDate(m['deletedAt'])),
          ),
        );
      }

      final notesList = data['notes'] as List<dynamic>? ?? const [];
      for (final n in notesList) {
        if (n is! Map) continue;
        final m = Map<String, dynamic>.from(n);
        final id = m['id'] as String?;
        final title = m['title'] as String? ?? '';
        final content = m['content'] as String? ?? '';
        final uid = m['userId'] as String?;
        if (id == null || uid == null) continue;
        await db.into(db.notes).insertOnConflictUpdate(
          NotesCompanion.insert(
            id: id,
            title: title,
            content: content,
            userId: uid,
            folderId: Value(m['folderId'] as String?),
            isPinned: Value(_coerceBool(m['isPinned'])),
            createdAt: Value(_parseDate(m['createdAt']) ?? DateTime.now()),
            updatedAt: Value(_parseDate(m['updatedAt']) ?? DateTime.now()),
            deletedAt: Value(_parseDate(m['deletedAt'])),
          ),
        );
        final plainContent = NoteUtils.extractLines(content).join(' ');
        await db.upsertNoteFts(id, title, plainContent, uid);
      }

      // 5. Apply tombstones.
      final deletedList = data['deleted'] as List<dynamic>? ?? const [];
      for (final d in deletedList) {
        if (d is! Map) continue;
        final m = Map<String, dynamic>.from(d);
        final entityId = m['entityId'] as String?;
        final entityType = m['entityType'] as String?;
        final deletedAt = _parseDate(m['deletedAt']);
        if (entityId == null || entityType == null || deletedAt == null) continue;

        if (entityType == 'note') {
          await (db.update(db.notes)..where((t) => t.id.equals(entityId))).write(
            NotesCompanion(deletedAt: Value(deletedAt)),
          );
          await db.deleteNoteFts(entityId);
        } else if (entityType == 'folder') {
          await (db.update(db.notes)..where((t) => t.folderId.equals(entityId))).write(
            const NotesCompanion(folderId: Value(null)),
          );
          await (db.delete(db.folders)..where((t) => t.id.equals(entityId))).go();
        }
      }
    } catch (e, stackTrace) {
      dev.log('Sync error', error: e, stackTrace: stackTrace);
      // Ops stay queued for the next attempt — we didn't ack them.
    }
  }

  /// How many local mutations are waiting to be pushed.
  Future<int> pendingOpCount() => _opDao.countPending();

  static DateTime? _parseDate(dynamic value) {
    if (value == null) return null;
    if (value is String) {
      try {
        return DateTime.parse(value);
      } catch (_) {
        return null;
      }
    }
    return null;
  }

  static bool _coerceBool(dynamic value) {
    if (value is bool) return value;
    if (value is num) return value != 0;
    if (value is String) return value.toLowerCase() == 'true';
    return false;
  }
}
